README.md
phpdevserver 0.5.9
==================

PHP Dev Server 集成了 Apache 2.4 , PHP 5.5.X/5.6.X/7.0.X/7.1.X , phpMyAdmin , win-Bash , 提供 PHP 開發者一個方便切換 PHP 版本所需的環境進行開發與測試(千萬別拿來當正式機提供服務)

PHP 本身會設定為 FASTCGI 的方式運作穩定性高。這個套件目前支援以下功能

- PHP 版本可以很簡單切換
- 所有集成的套件皆來自官方提供，沒有重新編譯過，保證純淨
- 沒有安裝檔，可以放到任何目錄使用然後執行一次 auto-config.bat 就可以完成所有設定
- Xdebug enabled ， 預設支援 Netbeans IDE
- zend-opcache enabled in Apache mode , disabled in CLI mode
- PHP 在 apache mode 和 cli mode 有切開不同的 module 載入
- php\_imagick 模組及轉檔命令檔(可執行 %MAGICK\_HOME%\convert.exe)
- PHP 7.x 目前狀況如下
  - php_mysql 正式被移除了，所以不會支援 mysql function，僅能使用 mysqli or PDO
- 只要找的到的官方 pecl dll 都可以自己加入。 



本套件沒有包進 MySQL，一方面是難包，一方面 MySQL 或 MariaDB 自己下載來安裝就很簡單了，而且更可彈性規劃資料目錄及使用特殊功能

## 環境需求 ##

- Windows 7 64bit 以上，如果有加入 Windows AD Server 應該是不能用
- [Microsoft Visual C++ 2008 Redistributable(x64)](https://www.microsoft.com/zh-tw/download/details.aspx?id=2092) : PHP 5.5 需要，系統若內建則不需要安裝
- [Microsoft Visual C++ 2012 Redistributable(x64)](https://www.microsoft.com/zh-TW/download/details.aspx?id=30679) : PHP 5.6 需要，系統若內建則不需要
- [Microsoft Visual C++ 2015 Redistributable(x64)](https://www.microsoft.com/en-us/download/details.aspx?id=48145) : PHP 7.0+ 需要，系統若內建則不需要

## 安裝 ##

- 將 phpdevserver 放到任何路徑，例如 C:\phpdevserver
- 執行 auto-config.bat 會自動設定好所有參數
- 請注意 : auto-config.bat 會修改系統變數 PATH 及 PHP\_INI\_SCAN\_DIR，並且會將 Apache 註冊為 Service
- 安裝完成後可以發現 Windows 右下角可以找到 ApacheMonitor 看看是否有變綠色代表 Apache 有成功運作
- 預設的 URL : http://localhost
- phpMyAdmin URL : http://localhost/phpmyadmin/

## 升級 ##

請依照以下步驟升級

- 先不要用 git 或覆蓋目錄的方式取代 phpdevserver
- 停止 Apache 服務
- 將 ApacheMonitor 退出
- 現在可以用 git 更新整個 phpdevserver 或直接下載回來取代
- 重新執行 auto-config.bat 就會自動升級並設定所有環境變數及網站設定檔
- 重啟 Apache Service


## 切換 PHP 版本 ##

在 phpdevserver/bin 下有以下批次檔，可以用來切換 php 版本，也會一併修改 Apache 設定

- switch-php55.bat
- switch-php56.bat
- switch-php70.bat
- switch-php71.bat

當切換成功後，必須重啟 Apache 使之生效，如果要讓 CLI 模式也生效，必須重新打開 DOS Console ，然後執行以下命令看看 PHP 版本有沒有切換成功

~~~
php -v
~~~

> 切換後可能發生的問題 :
> 
> session 可能讀取錯誤，如果原本是 PHP70 以上產生的 session file ，會和 PHP55~PHP56 的格式稍有不同，若發生此錯誤，刪除 session 檔案即可(**可自己寫 phpinfo() 查看 session 到底放甚麼地方**)，若是程式中是使用 db 存 session ，也是一樣將 session table 整個清掉才會正常。



## 反安裝 ##

- 先不要刪除 phpdevserver 目錄
- 先反註冊 Apache24 service , 以系統管理員執行 httpd -k uninstall
- 修改系統變數 PATH ，將 %PHPDEVSERVER_PATH% 拿掉
- 移除系統變數 PHPDEVSERVER_PATH
- 移除系統變數 PHPDEVSERVER\_PHP\_VERSION
- 移除系統變數 PHP\_INI\_SCAN_DIR
- 移除系統變數 MAGICK\_HOME
- 移除 ApacheMonitor 連結於 StartUP(以 Win10 為例可能會在 C:\ProgramData\Microsoft\Windows\Start Menu\Programs\StartUp[啟動])
- 現在可以刪除 phpdevserver 目錄了

## 性能問題 ##

由於預設有載入 xdebug，所以不管 cli 或 apache 模式都會比較慢，若覺得不需要 xdebug ，請自行將以下 xdebug 設定檔內載入的那一行註解即可

 - php[56~71]/conf.apache.d/11-xdebug.ini
 - php[56~71]/conf.cli.d/11-xdebug.ini


記得要重開 apache 服務才會生效



## 特別目錄說明 ##

- phpdevserver/Apache24/conf.d : php fcgid phpmyadmin 等設定都放這，建議 virtual host 或 alias 設定寫設定檔於此
- phpdevserver/phpXX/conf.apache.d : PHP 在 apache 運作會載入的模組設定都放在這
- phpdevserver/phpXX/conf.cli.d : PHP 在 CLI 模式載入的模組設定都放在

## PHP 預設已開啟功能 ##

- 模組設定區分 CLI 和 Apache 模式，分別於php路徑中 conf.cli.d 及 conf.apache.d 設定
- php.ini 預設以 php.ini-development 為主打開所有錯誤訊息方便偵錯
- 預設時區 date.timezone = UTF ，請參考 05-timezone.ini
- opcache : 64MB share memory enabled in apache mode , disabled in CLI mode
- bz2
- curl
- gd2
- imagick
- mysql + mysqli (注意 : PHP7.0 正式移除 mysql extension)
- mbstring
- openssl
- pdo_mysql + pdo_sqlite
- sockets
- xdebug
- 其他模組請參見 phpXX/ext 下的 dll，可自行新增 ini 於 conf.cli.d 或 conf.apache.d

若要修改任何模組載入，請盡量不要修改 php.ini，建議自行新增或修改 ini 於 PHPXX 中的 conf.apache.d 及 conf.cli.d


## 集成套件來源說明 ##

- Apache 2.4.25 64bit : From [http://www.apachehaus.com/](http://www.apachehaus.com/cgi-bin/download.plx)
- Mod FCGID 2.3.9 for Apache 2.4.x x64 : From [www.apachehaus.com](http://www.apachehaus.com/cgi-bin/download.plx)
- PHP 5.6.32 64bit : From [windows.php.net](http://windows.php.net/download/)
- PHP 5.5.38 64bit : From [windows.php.net](http://windows.php.net/download/)
- PHP 7.0.25 64bit : From [windows.php.net](http://windows.php.net/download/)
- PHP 7.1.11 64bit : From [windows.php.net](http://windows.php.net/download/)
- win-bash 1.1 : From [win-bash.sourceforge.net](http://win-bash.sourceforge.net/)
- phpMyAdmin 4.7.5 : From [www.phpmyadmin.net](https://www.phpmyadmin.net/)
- Imagick 6.9.3-7 64bit From [PECL](http://windows.php.net/downloads/pecl/deps/)
- php-xdebug 2.5.5 : From [xdebug.org](http://xdebug.org/)
- php-imagick 3.4.3 : From [PECL](http://windows.php.net/downloads/pecl/releases/imagick/)
- php-apcu 4.0.11(php 5.x)/5.1.8(php 7.x) : From [PECL](https://pecl.php.net/package/APCu)
- composer 1.5.2 : From [getcomposer.org](https://getcomposer.org )


CHANGELOG.md
phpdevserver Change Log
========================


0.5.10 2019-01-27
-----------------------
- 升級 phpmyadmin 至 4.8.5
- 升級 composer 至 1.8.0
- 增加 PHP 7.3.1 (未包含 imagick extension , 包含 xdebug 2.7.0BETA 1)


0.5.9 2017-11-13
-----------------------
- 升級 PHP 71 至 7.1.11
- 升級 PHP 70 至 7.0.25
- 升級 PHP 56 至 5.6.32
- 升級 phpmyadmin 至 4.7.5
- 升級 composer 至 1.5.2
- 升級 php-xdebug 至 2.5.5

0.5.8 2017-08-14
-----------------------
- 升級 PHP71 至 7.1.8
- 升級 PHP70 至 7.0.22
- 升級 PHP56 至 5.6.31
- 升級 phpmyadmin 至 4.7.3
- 升級 composer 至 1.5.1

0.5.7 2017-06-14
-----------------------
- 升級 PHP71 至 7.1.6
- 升級 PHP70 至 7.0.20
- 升級 php-xdebug 至 2.5.4
- 升級 composer 至 1.4.2
- 升級 phpmyadmin 至 4.7.1


0.5.6 2017-05-14
-----------------------
- 升級 PHP71 至 7.1.5
- 升級 PHP70 至 7.0.19
- 升級 php-xdebug 至 2.5.3


0.5.5 2017-04-16
-----------------------
- 升級 PHP71 至 7.1.4
- 升級 PHP70 至 7.0.18
- 升級 composer 至 1.4.1
- 升級 phpmyadmin 至 4.7.0
- 升級 php-imagick 至 3.4.3 正式版

0.5.4 2017-03-06
-----------------------
- 升級 composer 至 1.3.2
- 升級 phpmyadmin 至 4.6.6
- 升級 PHP70 至 7.0.16
- 升級 PHP71 至 7.1.2
- 升級 php-xdebug 至 2.5.1

0.5.3 2017-01-20
-----------------------

- 升級 PHP56 至 5.6.30
- 升級 PHP70 至 7.0.15
- 升級 PHP71 至 7.1.1
- 升級 php-imagick 至 3.4.3 RC2


0.5.2 2017-01-09
-----------------------

- 升級 phpmyadmin 至 4.6.5.2
- 升級 composer 至 1.3.1

0.5.1 2016-12-27
-----------------------

- 升級 composer 至 1.3.0
- 升級 PHP70 至 7.0.14
- 升級 PHP56 至 5.6.29
- 升級 Apache 至 2.4.25
  


0.5.0 2016-12-05
-----------------------

- 新增可切換 PHP71 : switch-php71
- **新增 PHP 7.1.0**
- **升級時請記得要先執行 auto-config才能讓 PHP7.1 各種環境設定好**
- 升級 composer 至 1.2.3
- 升級 php-xdebug 至 2.5.0 stable


0.4.5 2016-11-30
-----------------------

- 升級 phpMyAdmin 至 4.6.5.1
- 升級 php-xdebug 至 2.5.0 RC1

0.4.4 2016-11-14
-----------------------

- 升級 PHP70 至 7.0.13
- 升級 PHP56 至 5.6.28


0.4.3 2016-11-04
-----------------------

- 升級 composer 至 1.2.2
- 升級 PHP70 至 7.0.12
- 升級 PHP56 至 5.6.27


0.4.2 2016-10-07
-----------------------

- 升級 composer 至 1.2.1
- Apache24 缺少相關設定檔導致無法執行auto-config.bat
- PHP70 的 apcu 升級至 5.1.6
- 升級 PHP70 至 7.0.11
- 升級 PHP56 至 5.6.26



0.4.1 2016-08-22
-----------------------

- Apache24 來源改為 http://www.apachehaus.com/
- Apache24 升級至 2.4.23
- 升級 composer 至 1.2.0
- 升級 PHP70 至 7.0.10
- 升級 PHP56 至 5.6.25
- 升級 PHP55 至 5.5.38
- 升級 phpmyadmin 至 4.6.4
- 升級 php-xdebug 至 2.4.1

0.4.0 2016-07-12
-----------------------

- auto-config 加入 apache service uninstall 之後再 install
- 升級 composer 至 1.1.3
- 升級 PHP55 至 5.5.37
- 升級 PHP56 至 5.6.23
- 升級 PHP70 至 7.0.8
- 升級 PHP70 的 php-apcu 至 5.1.5
- 升級 PHP55/PHP56 的 php-apcu 至 4.0.11
- 升級 php-imagick 至 3.4.3 RC1
- 升級 phpMyAdmin 至 4.6.3
- 升級 Imagick 至 6.9.3-7
- 因 LICENSE 問題，Imagick 改為放完整的檔案(包含 include,lib)，並修改 MAGICK_HOME 環境變數
- 請務必執行一次 auto-config.bat 升級才能確保 MAGICK_HOME 是正確的

0.3.5 2016-03-29
-----------------------

- composer 升級至最新版
- 升級 phpMyAdmin 至 4.5.5.1
- 升級 PHP70 至 7.0.4
- 升級 PHP56 至 5.6.19
- 升級 PHP55 至 5.5.33
- PHP70/56/55 都加入 extension : php-apcu
- php-xdebug 升級至 2.4.0 正式版

0.3.4 2016-02-19
------------------

- composer 升級至最新版
- 升級 PHP70 至 7.0.3
- 升級 PHP56 至 5.6.18
- 升級 PHP55 至 5.5.32
- 升級 phpmyadmin 至 4.5.4.1
- 升級 php-xdebug 至 2.4.0 RC4

0.3.3 2016-01-14
------------------

- composer 升級至最新版
- 升級 PHP70 至 7.0.2
- 升級 PHP56 至 5.6.17
- 升級 PHP55 至 5.5.31
- 升級 phpmyadmin 至 4.5.3.1
- 升級 php-imagick 至 3.4.0RC5
- php_fileinfo 預設載入
- php_intl 預設載入

0.3.2 2015-12-13
------------------

- 升級 Apache 至 2.4.18
- 升級 php-imagick 至 3.4.0rc3
- PHP 7.0 加入 php-imagick 3.4.0rc3 (必須執行一次 auto-config.bat 會自動設定)
- 升級 xdebug 至 2.4.0 RC3
- composer 更新到最新版

0.3.1 2015-12-04
------------------

- 升級 PHP56 至 5.6.16
- 升級 PHP 7.0 至正式版
- php55/php56 升級 xdebug 至 2.4.0 RC2
- php70 增加 xdebug 2.4.0 RC2



0.3.0 2015-10-21
------------------

- 增加 PHP 7.0
  - php_mysql 正式被移除了，所以不會支援 mysql function，僅能使用 mysqli or PDO
  - imagick 目前不支援 , 待官方正式編譯好後才會支援
  - xdebug 目前不支援 , 待官方正式編譯好後才會支援
- 升級 PHP55 至 5.5.30
- 升級 PHP56 至 5.6.14
- 升級 phpMyAdmin 至 4.5.0.2
- 升級 composer
- 升級 Apache 至 2.4.17

0.2.5 2015-10-01
------------------

- 修改系統變數 Path 時，會保留原本的字串

0.2.4 2015-09-30
------------------

- 修復第一次執行 auto-config.bat ，php fastcgi 設定失敗的問題

0.2.3 2015-09-28
------------------

- 加入完整 ImageMagick 檔案，可以直接執行 convert 進行轉檔
- 修復 php imagick support format no value
- 修復 session 預設儲存於 C:\Windows 的問題，現在會存於環境變數所指定的位置，如 C:\Windows\Temp


0.2.2 2015-09-27
------------------

- 增加 composer

0.2.1 2015-09-26
------------------

- auto-config.bat 會自動將 ApacheMonitor 建立 StartUp 連結於開機時啟動
- 預設載入 php_sockets.dll

0.2.0 2015-09-26
------------------

- 大幅修改目錄結構是為了將來好升級，若有安裝 0.1.0 的人請重新安裝
- auto-config.bat 現在可以自動要求 Admin 權限
- 實作切換 PHP 5.6 及 5.5 機制，當未來有 PHP 7 的 Windows 版出來時才好升級
- 增加 php 5.5.29 nts 64bit
- 增加 php_imagick 3.3.0rc2 模組

0.1.0 2015-09-25
------------------

- 增加 php 5.6.13 nts 64bit
- 增加 xdebug 2.3.3
- 增加 phpMyAdmin 4.4.15
- 增加 Apache 2.4.16 64bit
- 增加 win-bash 1.1
- 增加 mod_fcgid-2.3.9-V14

phpmyadmin/libraries/language_stats.inc.php
<?php
/* Automatically generated file, do not edit! */
/* Generated by scripts/remove-incomplete-mo */

$GLOBALS["language_stats"] = array (
    'af' => 8,
    'ar' => 35,
    'az' => 51,
    'be' => 63,
    'be@latin' => 22,
    'bg' => 60,
    'bn' => 71,
    'br' => 16,
    'brx' => 0,
    'bs' => 11,
    'ca' => 98,
    'ckb' => 31,
    'cs' => 91,
    'cy' => 16,
    'da' => 92,
    'de' => 99,
    'el' => 98,
    'en_GB' => 76,
    'eo' => 13,
    'es' => 100,
    'et' => 99,
    'eu' => 14,
    'fa' => 25,
    'fi' => 67,
    'fil' => 0,
    'fr' => 99,
    'fy' => 18,
    'gl' => 82,
    'gu' => 0,
    'he' => 15,
    'hi' => 36,
    'hr' => 27,
    'hu' => 97,
    'hy' => 64,
    'ia' => 75,
    'id' => 68,
    'ig' => 2,
    'it' => 99,
    'ja' => 69,
    'ka' => 12,
    'kab' => 2,
    'kk' => 9,
    'km' => 5,
    'kn' => 3,
    'ko' => 80,
    'ksh' => 0,
    'ky' => 1,
    'li' => 1,
    'lt' => 41,
    'lv' => 17,
    'mk' => 16,
    'ml' => 1,
    'mn' => 18,
    'ms' => 10,
    'nb' => 56,
    'ne' => 1,
    'nl' => 100,
    'pa' => 2,
    'pl' => 81,
    'pt' => 53,
    'pt_BR' => 100,
    'ro' => 52,
    'ru' => 94,
    'si' => 58,
    'sk' => 80,
    'sl' => 100,
    'sq' => 95,
    'sr' => 21,
    'sr@latin' => 42,
    'sv' => 94,
    'ta' => 33,
    'te' => 8,
    'th' => 28,
    'tk' => 0,
    'tr' => 100,
    'tt' => 15,
    'ug' => 10,
    'uk' => 99,
    'ur' => 17,
    'uz' => 31,
    'uz@latin' => 30,
    'vi' => 82,
    'vls' => 2,
    'zh_CN' => 85,
    'zh_TW' => 98,
);
?>
